### ISSUE TITLE
Add -nameopt CANONICAL to "openssl x509"

### ISSUE THREAD

**Issue by michael-o** [openssl/openssl#11738]
Please add the name option `CANONICAL` (for subjects at least) which shall match the exact same value as used to calculate the hash. This is for debugging purposes like [here](https://stackoverflow.com/q/40723858/696632). Ideally, one could also request a DER dump of othis value via "openssl x509".

**Comment by richsalz** [comment]
But the hash is on the DER, so what should be output?

**Comment by michael-o** [comment]
That is correct, is on the correct form, but this could have spared me a lot of time. I tried to calculate the hash in Java noticed that:

* Java's X500Principal does not set UTF8STring (`0x0c`), but PrintableString
* It does not print, e.g. `2.5.4.5` value, as string, but as a `#` DER hex string sequence (TLV). Moreover, during canoncalization it does not lowercase the S/N value.

Having OpenSSL's output would have helped to find this way way faster.

**Comment by richsalz** [comment]
So you want to see a hexdump (maybe ff:22:3e:... format?)  That could be reasonable, adding a "der" format.  But what is canonical?  There *IS* no canonical string format defined anywhere.

**Comment by michael-o** [comment]
Yes, the hex dump could be either a hexstring or base64. The canonical format I am referring to is by `i2d_X509_NAME`. `X509_NAME_cmp` uses the canonical form to compare two names. This format I'd like to see.

RFC 2253 form:  `CN=Siemens Issuing CA Intranet Server 2017,OU=Siemens Trust Center,2.5.4.5=#13085a5a5a5a5a5a4237,O=Siemens,L=Muenchen,ST=Bayern,C=DE`
Canonical form: `cn=siemens issuing ca intranet server 2017,ou=siemens trust center,2.5.4.5=#13085a5a5a5a5a5a4237,o=siemens,l=muenchen,st=bayern,c=de`

At least for Java. in OpenSSL the RFC 2253 form is:
```
$ openssl x509 -in /usr/local/etc/ssl/certs/siemens-am.pem -subject  -noout -nameopt RFC2253            subject=CN=Siemens Issuing CA Internet Code Signing 2016,OU=Siemens Trust Center,serialNumber=ZZZZZZA8,O=Siemens,L=Muenchen,ST=Bayern,C=DE
```

**Comment by michael-o** [comment]
For the time being, I have patched OpenSSL in a very hacky way to get the bytes:

```diff
diff --git a/crypto/x509/x509_cmp.c b/crypto/x509/x509_cmp.c
index a7b90e6a42..dce4587e38 100644
--- a/crypto/x509/x509_cmp.c
+++ b/crypto/x509/x509_cmp.c
@@ -235,0 +236,6 @@ unsigned long X509_NAME_hash(X509_NAME *x)
+    int i;
+    for (i = 0; i < x->canon_enclen; i++) {
+      unsigned char c = (unsigned char) x->canon_enc[i];
+      printf("%02x", c);
+    }
+    printf("\n");
```

And the output:
```
$ .../bin/openssl x509 -hash -noout -in ca-cert.crt
310b300906035504060c0264653111300f06035504080c0862c3b672206c696e3112301006035504070c0962c39672206c20696e312e302c060355040a0c25d092d0b8d0bad0b8d0bfd0b5d0b4d0b8d18e20d092d0b8d0bad0b8d0bfd0b5d0b4d0b8d18e31163014060355040b0c0d6578616d706c6520636f72702e3128302606035504030c1f6d69636861656c2d6f20636572746966696361746520617574686f726974793124302206092a864886f70d0109010c156d69636861656c2d6f406578616d706c652e636f6d
5ba4b7de
```

**Comment by MrMabulous** [comment]
+1, this would be useful

**Comment by nhorman** [comment]
Sounds useful, but not likely to be planned, marking as community issue

**Comment by t8m** [comment]
Please note we are looking for community help implementing this issue. Please comment if you are interested in implementing this feature or if you have reasons why this feature is important enough so we should consider implementing it with higher priority than other requested features. The issue will be closed after 3 months otherwise.

**Comment by t8m** [comment]
This issue was closed because it was marked with the help wanted label and there was no interest in implementing this issue or any other activity for more than 2 months.
